core/primitives/strobe.hpp

```diff
@@ -0,0 +1,206 @@
+
+      const auto old_begin = begin_position_;
+      begin_position_ = current_position_ + 1;
+      current_state_ = kFlags;
+      absorb({old_begin, kFlags});
+
+      if constexpr (0 != (kFlags & (kFlag_C | kFlag_K))) {
+        if (current_position_ != 0) {
+          runF();
+        }
+      }
+    }
+
+    void runF() {
+      *as<uint8_t>(current_position_) ^= begin_position_;
+      *as<uint8_t>(current_position_ + 1) ^= 0x04;
+      *as<uint8_t>(kStrobeR + 1) ^= 0x80;
+      keccakf(as<uint64_t>());
+
+      current_position_ = 0;
+      begin_position_ = 0;
+    }
+
+    Strobe(const Strobe &) = delete;
+    Strobe &operator=(const Strobe &) = delete;
+
+    Strobe(Strobe &&) = delete;
+    Strobe &operator=(Strobe &&) = delete;
+
+   public:
+    Strobe()
+        : buffer_{reinterpret_cast<uint8_t *>(math::roundUp<kAlignment>(
+              reinterpret_cast<uintptr_t>(raw_data)))} {}
+
+    template <typename T, size_t N>
+    void initialize(const T (&label)[N]) {
+      constexpr bool kUseRuntimeCalculation = false;
+
+      if constexpr (kUseRuntimeCalculation) {
+        std::memset(as<uint8_t>(), 0, count<uint8_t>());
+        load<0ull>((uint8_t[6]){1, kStrobeR + 2, 1, 0, 1, 96});
+        load<6ull>("STROBEv1.0.2");
+        keccakf(as<uint64_t>());
+      } else {
+        load<0ull>((uint8_t[kBufferSize]){
+            0x9c, 0x6d, 0x16, 0x8f, 0xf8, 0xfd, 0x55, 0xda, 0x2a, 0xa7, 0x3c,
+            0x23, 0x55, 0x65, 0x35, 0x63, 0xdc, 0xc,  0x47, 0x5c, 0x55, 0x15,
+            0x26, 0xf6, 0x73, 0x3b, 0xea, 0x22, 0xf1, 0x6c, 0xb5, 0x7c, 0xd3,
+            0x1f, 0x68, 0x2e, 0x66, 0xe,  0xe9, 0x12, 0x82, 0x4a, 0x77, 0x22,
+            0x1,  0xee, 0x13, 0x94, 0x22, 0x6f, 0x4a, 0xfc, 0xb6, 0x2d, 0x33,
+            0x12, 0x93, 0xcc, 0x92, 0xe8, 0xa6, 0x24, 0xac, 0xf6, 0xe1, 0xb6,
+            0x0,  0x95, 0xe3, 0x22, 0xbb, 0xfb, 0xc8, 0x45, 0xe5, 0xb2, 0x69,
+            0x95, 0xfe, 0x7d, 0x7c, 0x84, 0x13, 0x74, 0xd1, 0xff, 0x58, 0x98,
+            0xc9, 0x2e, 0xe0, 0x63, 0x6b, 0x6,  0x72, 0x73, 0x21, 0xc9, 0x2a,
+            0x60, 0x39, 0x7,  0x3,  0x53, 0x49, 0xcc, 0xbb, 0x1b, 0x92, 0xb7,
+            0xb0, 0x5,  0x7e, 0x8f, 0xa8, 0x7f, 0xce, 0xbc, 0x7e, 0x88, 0x65,
+            0x6f, 0xcb, 0x45, 0xae, 0x4,  0xbc, 0x34, 0xca, 0xbe, 0xae, 0xbe,
+            0x79, 0xd9, 0x17, 0x50, 0xc0, 0xe8, 0xbf, 0x13, 0xb9, 0x66, 0x50,
+            0x4d, 0x13, 0x43, 0x59, 0x72, 0x65, 0xdd, 0x88, 0x65, 0xad, 0xf9,
+            0x14, 0x9,  0xcc, 0x9b, 0x20, 0xd5, 0xf4, 0x74, 0x44, 0x4,  0x1f,
+            0x97, 0xb6, 0x99, 0xdd, 0xfb, 0xde, 0xe9, 0x1e, 0xa8, 0x7b, 0xd0,
+            0x9b, 0xf8, 0xb0, 0x2d, 0xa7, 0x5a, 0x96, 0xe9, 0x47, 0xf0, 0x7f,
+            0x5b, 0x65, 0xbb, 0x4e, 0x6e, 0xfe, 0xfa, 0xa1, 0x6a, 0xbf, 0xd9,
+            0xfb, 0xf6});
+      }
+
+      current_position_ = 0;
+      current_state_ = kFlag_NU;
+      begin_position_ = 0;
+
+      metaAd<false>(label);
+    }
+
+    template <bool kMore, typename T, size_t N>
```

Comment:
Tbh this bool param is very cryptic when I see it in usages. Maybe it's better to replace it, for example, with enum class with meaningful constant names?

Reply:
This name for this param also uses in rust impl. The perpose of it is the only to check that you don't change the operation state between multiple calls.